flags.go

```diff
@@ -142,23 +185,113 @@ func (g *goforit) Enabled(ctx context.Context, name string) (enabled bool) {
 
 	g.flagsMtx.RLock()
 	defer g.flagsMtx.RUnlock()
+	enabled = false
 	if g.flags == nil {
-		enabled = false
+		log.Printf("[goforit] empty flags map")
 		return
 	}
 	lastRefreshTime = g.lastFlagRefreshTime
 	flag := g.flags[name]
 
-	// equality should be strict
-	// because Float64() can return 0
-	if f := g.rand(); f < flag.Rate {
-		enabled = true
+	if !flag.Active {
 		return
 	}
+
+	var mergedProperties = map[string]string{}
+
+	for k, v := range g.defaultTags {
+		mergedProperties[k] = v
+	}
+
+	for k, v := range properties {
+		mergedProperties[k] = v
+	}
+
+	for _, r := range flag.Rules {
+		res, err := r.Rule.Handle(flag.Name, mergedProperties)
+		if err != nil {
+			log.Printf("[goforit] error evaluating rule:\n", err)
+			return
+		}
+		var matchBehavior RuleAction
+		if res {
+			matchBehavior = r.OnMatch
+		} else {
+			matchBehavior = r.OnMiss
+		}
+		switch matchBehavior {
+		case "on":
+			enabled = true
+			return
+		case "off":
+			enabled = false
+			return
+		case "continue":
+			continue
+		default:
+			log.Printf("[goforit] unknown match behavior: " + string(matchBehavior))
+			return
+		}
+	}
 	enabled = false
 	return
 }
 
+//
+// func (g *goforit) Enabled(ctx, name, tags) bool {
+// 	tags = mergeTags(tags, g.defaultTags)
+// 	...
+// }
+
+func getProperty(props map[string]string, prop string) (string, error) {
+	if v, ok := props[prop]; ok {
+		return v, nil
+	} else {
+		return "", errors.New("No property " + prop + " in properties map or default tags.")
+	}
+}
+
+func (r *RateRule) Handle(flag string, props map[string]string) (bool, error) {
+	if r.Properties != nil {
+		// get the sha1 of the properties values concat
+		h := sha1.New()
+		// sort the properties for consistent behavior (should we sort on Refresh()?)
+		sort.Strings(r.Properties)
+		var buffer bytes.Buffer
+		buffer.WriteString(flag)
+		for _, val := range r.Properties {
+			prop, err := getProperty(props, val)
```

Comment:
We should add a separator between props, so that ("foo", "barblah") is not equivalent to ("foobar","blah"). Maybe use a zero byte, since that's unlikely to be in an actual property value.